pallets/gear/src/benchmarking/syscalls.rs

```diff
@@ -170,20 +176,27 @@ where
         Self::prepare_handle(module, 0)
     }
 
-    // TODO: add check for alloc and free result #2498
     pub fn free(r: u32) -> Result<Exec<T>, &'static str> {
         assert!(r <= max_pages::<T>() as u32);
 
         use Instruction::*;
         let mut instructions = vec![];
         for _ in 0..API_BENCHMARK_BATCH_SIZE {
             instructions.push(I32Const(r as i32));
-            instructions.push(Call(0));
-            instructions.push(Drop);
+            instructions.push(Call(0)); // alloc `r` pages
+            instructions.push(I32Const(i32::MAX));
+            instructions.push(I32Eq); // if alloc returns i32::MAX then it's error
+            instructions.push(If(BlockType::NoResult));
+            instructions.push(Unreachable);
+            instructions.push(End);
             for page in 0..r {
                 instructions.push(I32Const(page as i32));
-                instructions.push(Call(1));
-                instructions.push(Drop);
+                instructions.push(Call(1)); // free `page` page
+                instructions.push(I32Const(0));
+                instructions.push(I32Ne); // if free returns not 0 then it's error
+                instructions.push(If(BlockType::NoResult));
```

Comment:
Add separate function inside `code.rs` mod `body`, which append this construction:
```
If flag
    unreachable
end
```
to the given `instructions` vector. And use it here.